presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java

```diff
@@ -0,0 +1,225 @@
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.GenericLiteral;
+import com.facebook.presto.sql.tree.NullLiteral;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.QualifiedNameReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.metadata.FunctionKind.AGGREGATE;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step;
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static com.facebook.presto.sql.tree.ComparisonExpression.Type.GREATER_THAN_OR_EQUAL;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Converts INTERSECT query into UNION ALL..GROUP BY...WHERE
+ * Eg:  SELECT a FROM foo INTERSECT SELECT x FROM bar
+ * <p>
+ * =>
+ * <p>
+ * SELECT a
+ * FROM
+ * (SELECT a,
+ * COUNT(foo_marker) AS foo_cnt,
+ * COUNT(bar_marker) AS bar_cnt
+ * FROM
+ * (
+ * SELECT a, true as foo_marker, null as bar_marker FROM foo
+ * UNION ALL
+ * SELECT x, null as foo_marker, true as bar_marker FROM bar
+ * ) T1
+ * GROUP BY a) T2
+ * WHERE foo_cnt >= 1 AND bar_cnt >= 1;
+ */
+public class ImplementIntersectAsUnion
+        implements PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+        requireNonNull(symbolAllocator, "symbolAllocator is null");
+        requireNonNull(idAllocator, "idAllocator is null");
+
+        return SimplePlanRewriter.rewriteWith(new Rewriter(idAllocator, symbolAllocator), plan);
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private static final String INTERSECT_MARKER = "intersect_marker";
+        private final PlanNodeIdAllocator idAllocator;
+        private final SymbolAllocator symbolAllocator;
+
+        private Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitIntersect(IntersectNode node, RewriteContext<Void> rewriteContext)
```

Comment:
I played with the code a bit, and this is what I came up with. I haven't tested it, so it may have bugs:

```java
@Override
public PlanNode visitIntersect(IntersectNode node, RewriteContext<Void> rewriteContext)
{
    List<PlanNode> sources = node.getSources().stream()
            .map(rewriteContext::rewrite)
            .collect(Collectors.toList());

    List<Symbol> markers = allocateMarkers(sources.size());

    // identity projection for all the fields in each of the sources plus marker columns
    List<PlanNode> withMarkers = appendMarkers(markers, sources, node);

    // add a union over all the rewritten sources. The outputs of the union have the same name as the
    // original intersect node
    List<Symbol> outputs = node.getOutputSymbols();
    PlanNode union = union(withMarkers, ImmutableList.copyOf(Iterables.concat(outputs, markers)));

    // add count aggregations and filter rows where any of the counts is 0
    AggregationNode aggregation = computeCounts(union, outputs, markers);
    FilterNode filterNode = addFilter(aggregation);

    // drop the count columns
    return project(filterNode, outputs);
}

private List<Symbol> allocateMarkers(int count)
{
    ImmutableList.Builder<Symbol> markers = ImmutableList.builder();
    for (int i = 0; i < count; i++) {
        markers.add(symbolAllocator.newSymbol(INTERSECT_MARKER, BOOLEAN));
    }
    return markers.build();
}

private List<PlanNode> appendMarkers(List<Symbol> markers, List<PlanNode> nodes, IntersectNode intersect)
{
    ImmutableList.Builder<PlanNode> result = ImmutableList.builder();
    for (int i = 0; i < nodes.size(); i++) {
        result.add(appendMarkers(nodes.get(i), i, markers, intersect.sourceSymbolMap(i)));
    }
    return result.build();
}

private PlanNode appendMarkers(PlanNode node, int markerIndex, List<Symbol> markers, Map<Symbol, QualifiedNameReference> projections)
{
    ImmutableMap.Builder<Symbol, Expression> assignments = ImmutableMap.builder();
    assignments.putAll(projections);

    // add extra marker fields to the projection
    for (int i = 0; i < markers.size(); ++i) {
        Expression expression = (i == markerIndex) ? TRUE_LITERAL : new NullLiteral();
        assignments.put(markers.get(i), expression);
    }

    return new ProjectNode(idAllocator.getNextId(), node, assignments.build());
}

private PlanNode union(List<PlanNode> nodes, List<Symbol> outputs)
{
    ImmutableListMultimap.Builder<Symbol, Symbol> outputsToInputs = ImmutableListMultimap.builder();
    for (PlanNode source : nodes) {
        for (int i = 0; i < source.getOutputSymbols().size(); i++) {
            outputsToInputs.put(outputs.get(i), source.getOutputSymbols().get(i));
        }
    }

    return new UnionNode(idAllocator.getNextId(), nodes, outputsToInputs.build(), outputs);
}

private AggregationNode computeCounts(PlanNode node, List<Symbol> originalColumns, List<Symbol> markers)
{
    ImmutableMap.Builder<Symbol, FunctionCall> aggregations = ImmutableMap.builder();
    ImmutableMap.Builder<Symbol, Signature> signatures = ImmutableMap.builder();

    for (Symbol marker : markers) {
        Symbol output = symbolAllocator.newSymbol("count", BIGINT);

        aggregations.put(output, new FunctionCall(QualifiedName.of("count"), ImmutableList.of(marker.toQualifiedNameReference())));
        signatures.put(output, COUNT_AGGREGATION);
    }

    return new AggregationNode(idAllocator.getNextId(),
            node,
            originalColumns,
            aggregations.build(),
            signatures.build(),
            ImmutableMap.of(),
            ImmutableList.of(originalColumns),
            Step.SINGLE,
            Optional.empty(),
            1.0,
            Optional.empty());
}

private FilterNode addFilter(AggregationNode aggregation)
{
    List<Expression> predicates = aggregation.getAggregations().keySet().stream()
            .map(column -> new ComparisonExpression(GREATER_THAN_OR_EQUAL, column.toQualifiedNameReference(), new GenericLiteral("BIGINT", "1")))
            .collect(toImmutableList());

    return new FilterNode(idAllocator.getNextId(), aggregation, ExpressionUtils.and(predicates));
}

private PlanNode project(PlanNode node, List<Symbol> columns)
{
    return new ProjectNode(
            idAllocator.getNextId(),
            node,
            columns.stream().collect(Collectors.toMap(Function.identity(), Symbol::toQualifiedNameReference)));
}
```